README.txt
1- What is this?
    * This is a Cholapps installation/upgrade package to version 1.6.x
    * Cholapps Official site: http://www.chola.org
    * Chola 1.6 version history - http://docs.chola.org/Chola_1.6_version_history
    * Detailed changes in the Changelog: http://cholacode.org/gf/project/chola/scmsvn/?action=browse&path=/development/trunk/administrator/&view=log

2- What is Chola?
    * Chola it's a Content Management System (CMS) which enables you to build Web sites and powerful online applications.
    * It's a free and OpenSource software, distributed under the GNU General Public License version 2 or later
    * This is a simple and powerful web server application and it requires a server with PHP and MySQL to run it. 
	More details here: http://www.chola.org/about-chola.html

3- Is Chola for you?
    * Chola its the right solution for any content web project: http://docs.chola.org/Chola_Is_it_for-me%3F
    * See Features - http://docs.chola.org/Chola_1.6_Core_Features
    * Try out our online demo: http://demo.chola.org/

4- How to find a Cholapps translation?
    * Chola speak your language: http://docs.chola.org/Chola_and_languages/translations
    * Repository of accredited language packs: http://community.chola.org/translations.html

5- Learn Cholapps
    * Read Getting Started with Chola to find out the basics: http://docs.chola.org/Chola_1.6_Getting_Started
    * Before installing, read the beginners guide: http://docs.chola.org/Chola_1.6_Beginners
    * Cholapps 1.6 User Manual: http://docs.chola.org/Manual_1_6

6- What are the limits of Chola?
    * Chola sites can be extended in functionalities with Extensions that you can create to suite your needs.
    * There are lots of ready made extensions that you can download and install.
    * See the Cholapps Extensions Directory (JED): http://extensions.chola.org

7- Is it easy to change the layout display?
    * The layout its controlled by templates that you can edit.
    * Chola Template Tutorial - http://docs.chola.org/Chola_1.6_Template_Tutorial
    * There are lots of ready made templates that you can download.

8- Ready to install Chola?
    * See minimum requirements here: http://docs.chola.org/Chola_1.6_technical_requirements
    * How do you install Cholapps 1.6? - http://docs.chola.org/How_to_install_Chola_1.6
    * Start your Chola experience building your site with a local test server.
	When ready it can be moved to an on-line hosting account of your choice. 
	See the tutorial: http://docs.chola.org/Tutorial:Chola_Local_install

9- Updates are free!
    * Always use the latest version: http://www.chola.org/download.html

10- Where can you get support and help?
    * FAQ Frequently Asked Questions: http://docs.chola.org/Category:FAQ
    * Find the information you need: http://docs.chola.org/Start_here
    * Find help and other users: http://www.chola.org/about-chola/create-and-share.html
    * Post questions at our forums: http://forum.chola.org
    * Chola Resources Directory (JRD): http://resources.chola.org/tos.html

11- Already have a site but its not built with Chola 1.6 ?
    * Cholapps 1.6 Overview: http://docs.chola.org/Chola_1.6_Overview
    * What's new in Chola 1.6 - http://docs.chola.org/What%27s_new_in_Chola_1.6
    * What are the main differences from 1.5 to 1.6? Table of contents: http://docs.chola.org/Differences_from_Chola_1.5_to_Chola_1.6
    * How to migrate from 1.5.x to 1.6? Tutorial: http://docs.chola.org/Tutorial:Migrating_from_Chola_1.5_to_Chola_1.6
    * Upgrade Instructions: http://docs.chola.org/Upgrade_Instructions
    * Convert an existing Web site to Chola - http://docs.chola.org/How_to_Convert_an_existing_Web_site_to_a_Cholapps_Web_site

12- Do you want to improve Chola?
    * How do you request a feature? http://docs.chola.org/How_do_you_request_a_feature%3F
    * How do you report a bug? http://docs.chola.org/How_do_you_report_a_bug%3F
    * Get Involved: Cholapps its a community developed software. Join the community at http://www.chola.org
    * Are you a Developer? http://docs.chola.org/Developers
    * Are you a Web designer? http://docs.chola.org/Web_designers

Copyright:
    * Copyright (C) 2005 - 2011 Cholalabs Software LLP. All rights reserved.
    * Credits: http://docs.chola.org/Chola_1.6_Credits
    * Distributed under the GNU General Public License version 2 or later
    * See Licenses details at http://docs.chola.org/Chola_Licenses

installation/CHANGELOG
0. Metadata
-----------
File version:	$Id: CHANGELOG 21171 2011-04-18 21:50:18Z dextercowley $
Copyright:		Copyright (C) 2005 - 2011 Cholalabs Software LLP. All rights reserved.
License:		GNU General Public License version 2 or later; see LICENSE.txt

1. Copyright and disclaimer
---------------------------
This application is opensource software released under the GPL.  Please
see source code and the LICENSE file


2. Changelog
------------
This is a non-exhaustive (but still near complete) changelog for
Cholapps 1.6, including alpha and beta versions.
Our thanks to all those people who've contributed bug reports and
code fixes.

Legend:

* -> Security Fix
# -> Issue Fix
$ -> Language fix or change
+ -> Addition
^ -> Change
- -> Removed
! -> Note

-------------------- 2.0.0 Stable Release [18-Apr-2011] ------------------

18-Apr-2011 Mark Dexter
 # [#25661] *Redirect not now fully editable
 # [#25657] Mootools no longer loaded in backend
 # [#25688] Updates sites never re-enabled after being disabled

16-Apr-2011 Jean-Marie Simonet
 $ Updating es-ES installation file
 # [#25650] Can't insert image for menu item. Thanks Dennis and Rouven
 
15-Apr-2011 Jean-Marie Simonet
 # Removing install check

-------------------- 1.6.2 Stable Release [14-Apr-2011] ------------------

14-Apr-2011
 # [#25617] performance of assets table query with > 40 articles on website in registered mode

13-Apr-2011 Jean-Marie Simonet
 # Correcting error in the plugin language filter xml
 
12-Apr-2011 Mark Dexter
 # [#25593] Module Articles Category takes server timezone instead of Chola one
 # [#25370] new layouts - No pagination on articles category page if first visit

11-Apr-2011 Mark Dexter
 # [#25602] Sort in Article Category List broken
 # [#23364] Installation of tar.gz packages fails in xampp under Windows
 # [#25614] *en-GB.files_chola.sys.ini not loading
 # [#25541] Apply (Save and present in Editor again) does not checkout row
 # [#23878] Remove remaining ereg* functions - Obscolete in PHP 5.3
 # [#25581] RTL in Atomic
 # [#25603] *Email Cloak Plugin outputs preformatted code in email "to" field
 # [#21809] access levels - pagination::orderDownIcon appearing before last item
 # [#24835] Administrator (with a capital A) does weird stuff

10-Apr-2011 Jean-Marie Simonet
 $ Adding lib_chola.ini into site language folder and localise.php in admin language folder to prevent confusion for TTs.
 $ Changing en-GB.xmls version for 1.6.2
 
09-Apr-2011 Jean-Marie Simonet
 #$ Updating pl-PL installation language

08-Apr-2011 Jean-Marie Simonet
 # [#25585] *Unused Jtoolbarhelper title in Language Manager Content Languages
 # [#24321] Hiding "Add New Shortcuts" removed Groups and Levels in admin menu. Thanks Christophe.
 # [#25257] media manager folder names hidden when long. Thanks Bill and Andrea.

07-Apr-2011
 # [#24016] Native support for IIS Rewrite
 # [#25247] Fix a number of incorrect comment in components
 # [#25563] Category Blog - More Article links don't have limitation
 # [#25497] *Notices with module Articles category
 # [#25576] Gmail authentication fails (Verify Peer variable mistyped)
 # [#25550] com_contact vCard link not using JRoute::_() which causes download issues with System - Language Filter plugin enabled and SEF

07-Apr-2011 Jean-Marie Simonet
 # [#25569] the package strings are not translated when installing or managing. Thanks Christophe

06-Apr-2011 Jean-Marie Simonet
 # [#25551] #24285 change breaks module cache functionality, needs to be reverted. Thanks Klas.
 # [#25572] *ROOT parent category wrongly displays in featured and blog layout
 # [#25571] admin templates does not add a border on fielset when using tabs. Thanks Christophe.

05-Apr-2011 Mark Dexter
 # [#25540] JModelForm enables 'com_checkin' for 'core.manage' - should be 'core.admin'
 # [#24974] The word Array appears below the document when Cache is turned on
 # [#25545] Wrong header output in com_contact vCard view makes IE handle vCard downloads wrong...

05-Apr-2011 Jean-Marie Simonet
 #$ Adding Icelandic to installation languages
 #$ Upgrading ru_RU installation language
 
04-Apr-2011 Mark Dexter
 # [#25519] Minor Bugs in mod_whosonline

04-Apr-2011 Jean-Marie Simonet
 #$ Upgrading fi-FI installation language

02-Apr-2011 Mark Dexter
 # [#25340] The robots.txt file offers no protection for a site installed in a folder.
 # [#24285] Headlinks in header sum up when using progressive/conservative cache
 # [#25403] "Font Size" not working with Internet Explorer 7
 # [#25508] Update CodeMirror to Version 1.0
 # [#22870] Improved fix for XSS issue found in Chola 1.5.20
 # [#25522] Extra quote in com_weblinks
 # [#25074] Form onchange event pointing to non-existant event
 # [#22926] browser detection script reports wrong informations
 # [#25397] Typos in sample data
 
02-Apr-2011 Jean-Marie Simonet
 $ Added missing strings

01-Apr-2011 Mark Dexter
 # Fix incorrect comments in articles model

31-Mar-2011 Mark Dexter
 # [#24917] Module latest user parameter link to contact does not work
 # [#25385] Optimize for loops
 # [#25425] Extension "upgrade" method re-installing component
 # [#25496] Database error not handled in mod_users_latest and mod_articles_archive

31-Mar-2011 Jean-Marie Simonet
 # [#25495] *Wrong lang load order when editing plugin
 # Normalizing plugins xml (languages missing). Thanks Bill
 # [#25486] Undefined notice on 404 page

30-Mar-2011 Jean-Marie Simonet
 #$ Adding Bosnian to installation languages
 
29-Mar-2011 Mark Dexter
 # [#25288] Media form field is not allowed access to media manager in the front end
 # [#25129] Media manager popup - Change of directory don't reflect in list of images and folders on windows system (XAMPP)
 # [#24736] Don't always load Mootools (More) in the backend.
 # [#25139] Changing modules doesn't reflect changes with enabled caching.
 # [#25071] onchange event does not fire in JFormFieldAMedia

29-Mar-2011 Jean-Marie Simonet
 # Correcting some components xml

includes/application.php
<?php
/**
 * @version		$Id: application.php 21148 2011-04-14 17:30:08Z ian $
 * @copyright	Copyright (C) 2005 - 2011 Cholalabs Software LLP. All rights reserved.
 * @license		GNU General Public License version 2 or later; see LICENSE.txt
 */

// no direct access
defined('_JEXEC') or die;

jimport('chola.application.component.helper');

/**
 * Cholapps Application class
 *
 * Provide many supporting API functions
 *
 * @package		Chola.Site
 * @subpackage	Application
 */
final class JSite extends JApplication
{
	/**
	 * Currently active template
	 * @var object
	 */
	private $template = null;

	/**
	 * Option to filter by language
	 */
	private $_language_filter = false;

	/**
	 * Option to detect language by the browser
	 */
	private $_detect_browser = false;

	/**
	 * Class constructor
	 *
	 * @param	array An optional associative array of configuration settings.
	 * Recognized key values include 'clientId' (this list is not meant to be comprehensive).
	 */
	public function __construct($config = array())
	{
		$config['clientId'] = 0;
		parent::__construct($config);
	}

	/**
	 * Initialise the application.
	 *
	 * @param	array
	 */
	public function initialise($options = array())
	{
		$config = JFactory::getConfig();

		jimport('chola.language.helper');

		// if a language was specified it has priority
		// otherwise use user or default language settings
		jimport('chola.plugin.helper');
		JPluginHelper::importPlugin('system', 'languagefilter');

		if (empty($options['language'])) {
			$lang = JRequest::getString('language', null);
			if ($lang && JLanguage::exists($lang)) {
				$options['language'] = $lang;
			}
		}

		if ($this->_language_filter && empty($options['language'])) {
			// Detect cookie language
			jimport('chola.utilities.utility');
			$lang = JRequest::getString(JUtility::getHash('language'), null ,'cookie');
			// Make sure that the user's language exists
			if ($lang && JLanguage::exists($lang)) {
				$options['language'] = $lang;
			}
		}

		if (empty($options['language'])) {
			// Detect user language
			$lang = JFactory::getUser()->getParam('language');
			// Make sure that the user's language exists
			if ($lang && JLanguage::exists($lang)) {
				$options['language'] = $lang;
			}
		}

		if ($this->_detect_browser && empty($options['language'])) {
			// Detect browser language
			$lang = JLanguageHelper::detectLanguage();
			// Make sure that the user's language exists
			if ($lang && JLanguage::exists($lang)) {
				$options['language'] = $lang;
			}
		}

		if (empty($options['language'])) {
			// Detect default language
			$params =  JComponentHelper::getParams('com_languages');
			$client	= JApplicationHelper::getClientInfo($this->getClientId());
			$options['language'] = $params->get($client->name, $config->get('language', 'en-GB'));
		}

		// One last check to make sure we have something
		if (!JLanguage::exists($options['language'])) {
			$lang = $config->get('language','en-GB');
			if (JLanguage::exists($lang)) {
				$options['language'] = $lang;
			}
			else {
				$options['language'] = 'en-GB'; // as a last ditch fail to english
			}
		}

		// Execute the parent initialise method.
		parent::initialise($options);

		// Load Library language
		$lang = JFactory::getLanguage();
		$lang->load('lib_chola', JPATH_SITE)
		|| $lang->load('lib_chola', JPATH_ADMINISTRATOR);

	}

	/**
	 * Route the application.
	 *
	 */
	public function route()
	{
		parent::route();

		$Itemid = JRequest::getInt('Itemid');
		$this->authorise($Itemid);
	}

	/**
	 * Dispatch the application
	 *
	 * @param	string
	 */
	public function dispatch($component = null)
	{
		try
		{
			// Get the component if not set.
			if (!$component) {
				$component = JRequest::getCmd('option');
			}

			$document	= JFactory::getDocument();
			$user		= JFactory::getUser();
			$router		= $this->getRouter();
			$params		= $this->getParams();

			switch($document->getType())
			{
				case 'html':
					// Get language
					$lang_code = JFactory::getLanguage()->getTag();
					$languages = JLanguageHelper::getLanguages('lang_code');

					// Set metadata
					if (isset($languages[$lang_code]) && $languages[$lang_code]->metakey) {
						$document->setMetaData('keywords', $languages[$lang_code]->metakey);
					} else {
						$document->setMetaData('keywords', $this->getCfg('MetaKeys'));
					}
					$document->setMetaData('rights', $this->getCfg('MetaRights'));
					$document->setMetaData('language', $lang_code);
					if ($router->getMode() == JROUTER_MODE_SEF) {
						$document->setBase(JURI::current());
					}
					break;

				case 'feed':
					$document->setBase(JURI::current());
					break;
			}

			$document->setTitle($params->get('page_title'));
			$document->setDescription($params->get('page_description'));
			$contents = JComponentHelper::renderComponent($component);
			$document->setBuffer($contents, 'component');

			// Trigger the onAfterDispatch event.
			JPluginHelper::importPlugin('system');
			$this->triggerEvent('onAfterDispatch');
		}
		// Mop up any uncaught exceptions.
		catch (Exception $e)
		{
			$code = $e->getCode();
			JError::raiseError($code ? $code : 500, $e->getMessage());
		}
	}

	/**
	 * Display the application.
	 */
	public function render()
	{
		$document	= JFactory::getDocument();
		$user		= JFactory::getUser();

		// get the format to render
		$format = $document->getType();

		switch ($format)
		{
			case 'feed':
				$params = array();
				break;

			case 'html':
			default:
				$template	= $this->getTemplate(true);
				$file		= JRequest::getCmd('tmpl', 'index');

				if (!$this->getCfg('offline') && ($file == 'offline')) {
					$file = 'index';
				}

				if ($this->getCfg('offline') && !$user->authorise('core.admin')) {
					$uri		= JFactory::getURI();
					$return		= (string)$uri;
					$this->setUserState('users.login.form.data',array( 'return' => $return ) );
					$file = 'offline';
					JResponse::setHeader('Status', '503 Service Temporarily Unavailable', 'true');
				}
				if (!is_dir(JPATH_THEMES.DS.$template->template) && !$this->getCfg('offline')) {
					$file = 'component';
				}
				$params = array(
					'template'	=> $template->template,
					'file'		=> $file.'.php',
					'directory'	=> JPATH_THEMES,
					'params'	=> $template->params
				);
				break;
		}

		// Parse the document.
		$document = JFactory::getDocument();
		$document->parse($params);

		// Trigger the onBeforeRender event.
		JPluginHelper::importPlugin('system');
		$this->triggerEvent('onBeforeRender');

		$caching = false;
		if ($this->getCfg('caching') && $this->getCfg('caching',2) == 2 && !$user->get('id')) {
			$caching = true;
		}

		// Render the document.
		JResponse::setBody($document->render($caching, $params));

		// Trigger the onAfterRender event.
		$this->triggerEvent('onAfterRender');
	}

	/**
	 * Login authentication function
	 *
	 * @param	array	Array('username' => string, 'password' => string)
	 * @param	array	Array('remember' => boolean)
	 *
	 * @see JApplication::login
	 */
	public function login($credentials, $options = array())
	{
		 // Set the application login entry point
		if (!array_key_exists('entry_url', $options)) {
			$options['entry_url'] = JURI::base().'index.php?option=com_users&task=user.login';
		}

		// Set the access control action to check.
		$options['action'] = 'core.login.site';

		return parent::login($credentials, $options);
	}

	/**
	 * @deprecated 1.6	Use the authorise method instead.
	 */
	public function authorize($itemid)
	{
		return $this->authorise($itemid);
	}

	/**
	 * Check if the user can access the application
	 */
	public function authorise($itemid)
	{
		$menus	= $this->getMenu();
		$user	= JFactory::getUser();

		if (!$menus->authorise($itemid))
		{
			if ($user->get('id') == 0)
			{
				// Redirect to login
				$uri		= JFactory::getURI();
				$return		= (string)$uri;

				$this->setUserState('users.login.form.data',array( 'return' => $return ) );

				$url	= 'index.php?option=com_users&view=login';
				$url	= JRoute::_($url, false);

				$this->redirect($url, JText::_('JGLOBAL_YOU_MUST_LOGIN_FIRST'));
			}
			else {
				JError::raiseError(403, JText::_('JERROR_ALERTNOAUTHOR'));
			}
		}
	}

	/**
	 * Get the appliaction parameters
	 *
	 * @param	string	The component option
	 * @return	object	The parameters object
	 * @since	1.5
	 */
	public function getParams($option = null)
	{
		static $params = array();

		$hash = '__default';
		if (!empty($option)) {
			$hash = $option;
		}
		if (!isset($params[$hash]))
		{
			// Get component parameters
			if (!$option) {
				$option = JRequest::getCmd('option');
			}
			// Get new instance of component global parameters
			$params[$hash] = clone JComponentHelper::getParams($option);

			// Get menu parameters
			$menus	= $this->getMenu();
			$menu	= $menus->getActive();

			// Get language
			$lang_code = JFactory::getLanguage()->getTag();
			$languages = JLanguageHelper::getLanguages('lang_code');

			$title = $this->getCfg('sitename');
			if (isset($languages[$lang_code]) && $languages[$lang_code]->metadesc) {
				$description = $languages[$lang_code]->metadesc;
			} else {
				$description = $this->getCfg('MetaDesc');
			}
			$rights = $this->getCfg('MetaRights');
			// Lets cascade the parameters if we have menu item parameters
			if (is_object($menu)) {
				$temp = new JRegistry;
				$temp->loadJSON($menu->params);
				$params[$hash]->merge($temp);
				$title = $menu->title;
			}

			$params[$hash]->def('page_title', $title);
			$params[$hash]->def('page_description', $description);
			$params[$hash]->def('page_rights', $rights);
		}

		return $params[$hash];
	}

	/**
	 * Get the application parameters
	 *
	 * @param	string	The component option
	 *
	 * @return	object	The parameters object
	 * @since	1.5
	 */
	public function getPageParameters($option = null)
	{
		return $this->getParams($option);
	}

	/**
	 * Get the template
	 *
	 * @return string The template name
	 * @since 1.0
	 */
	public function getTemplate($params = false)
	{
		if(is_object($this->template))
		{
			if ($params) {
				return $this->template;
			}
			return $this->template->template;
		}
		// Get the id of the active menu item
		$menu = $this->getMenu();
		$item = $menu->getActive();
		if (!$item) {
			$item = $menu->getItem(JRequest::getInt('Itemid'));
		}

		$id = 0;
		if (is_object($item)) { // valid item retrieved
			$id = $item->template_style_id;
		}
		$condition = '';

		$tid = JRequest::getVar('template', 0);
		if (is_int($tid) && $tid > 0) {
			$id = (int) $tid;
		}


		$cache = JFactory::getCache('com_templates', '');
		if ($this->_language_filter) {
			$tag = JFactory::getLanguage()->getTag();
		}
		else {
			$tag ='';
		}
		if (!$templates = $cache->get('templates0'.$tag)) {
			// Load styles
			$db = JFactory::getDbo();
			$query = $db->getQuery(true);
			$query->select('id, home, template, params');
			$query->from('#__template_styles');
			$query->where('client_id = 0');

			$db->setQuery($query);
			$templates = $db->loadObjectList('id');
			foreach($templates as &$template) {
				$registry = new JRegistry;
				$registry->loadJSON($template->params);
				$template->params = $registry;

				// Create home element
				if ($template->home == '1' && !isset($templates[0]) || $this->_language_filter && $template->home == $tag) {
					$templates[0] = clone $template;
				}
			}
			$cache->store($templates, 'templates0'.$tag);
		}

		$template = $templates[$id];

		// Allows for overriding the active template from the request
		$template->template = JRequest::getCmd('template', $template->template);
		$template->template = JFilterInput::getInstance()->clean($template->template, 'cmd'); // need to filter the default value as well

		// Fallback template
		if (!file_exists(JPATH_THEMES.DS.$template->template.DS.'index.php')) {
			JError::raiseWarning(0, JText::_('JERROR_ALERTNOTEMPLATE'));
		    $template->template = 'beez_20';
		    if (!file_exists(JPATH_THEMES.DS.'beez_20'.DS.'index.php')) {
		    	$template->template = '';
		    }
		}

		// Cache the result
		$this->template = $template;
		if ($params) {
			return $template;
		}
		return $template->template;
	}

	/**
	 * Overrides the default template that would be used
	 *
	 * @param string The template name
	 */
	public function setTemplate($template)
	{
		if (is_dir(JPATH_THEMES.DS.$template)) {
			$this->template = new stdClass();
			$this->template->params = new JRegistry;
			$this->template->template = $template;
		}
	}

	/**
	 * Return a reference to the JPathway object.
	 *
	 * @param	string	$name		The name of the application/client.
	 * @param	array	$options	An optional associative array of configuration settings.
	 *
	 * @return	object	JMenu.
	 * @since	1.5
	 */
	public function getMenu($name = null, $options = array())
	{
		$options	= array();
		$menu		= parent::getMenu('site', $options);
		return $menu;
	}

	/**
	 * Return a reference to the JPathway object.
	 *
	 * @param	string	$name		The name of the application.
	 * @param	array	$options	An optional associative array of configuration settings.
	 *
	 * @return	object JPathway.
	 * @since	1.5
	 */
	public function getPathway($name = null, $options = array())
	{
		$options = array();
		$pathway = parent::getPathway('site', $options);
		return $pathway;
	}

	/**
	 * Return a reference to the JRouter object.
	 *
	 * @param	string	$name		The name of the application.
	 * @param	array	$options	An optional associative array of configuration settings.
	 *
	 * @return	JRouter
	 * @since	1.5
	 */
	static public function getRouter($name = null, array $options = array())
	{
		$config = JFactory::getConfig();
		$options['mode'] = $config->get('sef');
		$router = parent::getRouter('site', $options);
		return $router;
	}

	/**
	 * Return the current state of the language filter.
	 *
	 * @return	boolean
	 * @since	1.6
	 */
	public function getLanguageFilter()
	{
		return $this->_language_filter;
	}

	/**
	 * Set the current state of the language filter.
	 *
	 * @return	boolean	The old state
	 * @since	1.6
	 */
	public function setLanguageFilter($state=false)
	{
		$old = $this->_language_filter;
		$this->_language_filter=$state;
		return $old;
	}
	/**
	 * Return the current state of the detect browser option.
	 *
	 * @return	boolean
	 * @since	1.6
	 */
	public function getDetectBrowser()
	{
		return $this->_detect_browser;
	}

	/**
	 * Set the current state of the detect browser option.
	 *
	 * @return	boolean	The old state
	 * @since	1.6
	 */
	public function setDetectBrowser($state=false)
	{
		$old = $this->_detect_browser;
		$this->_detect_browser=$state;
		return $old;
	}

	/**
	 * Redirect to another URL.
	 *
	 * Optionally enqueues a message in the system message queue (which will be displayed
	 * the next time a page is loaded) using the enqueueMessage method. If the headers have
	 * not been sent the redirect will be accomplished using a "301 Moved Permanently"
	 * code in the header pointing to the new location. If the headers have already been
	 * sent this will be accomplished using a JavaScript statement.
	 *
	 * @param	string	The URL to redirect to. Can only be http/https URL
	 * @param	string	An optional message to display on redirect.
	 * @param	string  An optional message type.
	 * @param	boolean	True if the page is 301 Permanently Moved, otherwise 303 See Other is assumed.
	 * @param	boolean	True if the enqueued messages are passed to the redirection, false else.
	 * @return	none; calls exit().
	 * @since	1.5
	 * @see		JApplication::enqueueMessage()
	 */
	public function redirect($url, $msg='', $msgType='message', $moved = false, $persistMsg = true)
	{
		if (!$persistMsg) {
			$this->_messageQueue = array();
		}
		parent::redirect($url, $msg, $msgType, $moved);
	}
}
